### PR TITLE
Added blueprint EP038

### PR DESCRIPTION
**Pull request by Alopalao**
Related to `coloring` issue [#45](https://github.com/kytos-ng/coloring/issues/45)

### Summary

Blueprint detailing vulnerabilities that `coloring` presents and how to solve them.

### CODE REVIEW

File: docs/blueprints/EP038.rst
Comment (Alopalao) [inline docs/blueprints/EP038.rst:61]: Priority for `of_lldp` flows is related to its issue [#85](https://github.com/kytos-ng/of_lldp/issues/85)

File: docs/blueprints/EP038.rst
Comment (viniarck) [inline docs/blueprints/EP038.rst:38]: Do we really need this extra flow with `dl_src` and `actions=drop`? Based on OF1.3 spec:

> If the table-miss flow entry does not exist, by default packets unmatched by flow entries are dropped
(discarded). A switch configuration, for example using the OpenFlow Configuration Protocol, may override
this default and specify another behaviour.

Since your proposal is narrowing down the match to include `in_port` then, a matched `dl_src` from a `in_port` that doesn't have a flow rule won't match. Unless network engineers install a flow table miss, but that's outside of `coloring` responsibilities, so looks like we could get away without having to maintain this extra drop flow, right? Unless the intention is to also try to optimize the lookup when it gets dropped, but usually switches have internal optimization for their lookups, so not sure if it's worth the effort to maintain and measure this efficiency. Food for thought.

Comment (viniarck) [reply docs/blueprints/EP038.rst:38]: Another related question: How will it detect if it's coming from "inside the topology"? Currently, `coloring` is trusting `of_lldp` so any discovered link will be consider to reach to an adjacent switch, and consequently that interface will have `nni: True`. 

So, maybe we should make it explicit that `of_lldp` shouldn't be enabled on interfaces where it doesn't expect to install `coloring` related flows? That would need confirmation if it'd be acceptable, if we need a higher level of control then maybe we might need a new API on `coloring` to have more control on which interfaces should be enabled.

**Review by viniarck — COMMENTED** [review]
@Alopalao congrats on leading this blueprint with @italovalcy. 

I raised some questions for you to think about and research. Also, depending on the approach we'll go for we might also need a migration script for network operators upgrading to `2023.2`. 

Speaking of `coloring` priority, it might be worth also take the opportunity to extract it in a settings variable, but preserving the same 50k value, same pattern that we have on other NApps that push flow. Wdyt?

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:38]: > How will it detect if it's coming from "inside the topology"?

Oh I meant that we could use the information that `coloring` has about link when creating flows. Link dictionaries contain the endpoints and the ports they are connected to. The idea is to use these ports for the matching field `in_port`.

Comment (viniarck) [reply docs/blueprints/EP038.rst:38]: I see. At is is though, If this customer port is connected to a switch, and if it sends LLDP packets compatible with the flow entry that `of_lldp` installs then it would also end up as a link in our topology. Although if it's not a switch managed by kytos, then `kytos/of_lldp.interface.is.nni` wouldn't get discovered. 

It'd be great to clarify what we're meaning by customer connection in this case, is it also a switch managed by our controler or not? or can it be both? Depending on the scope we might need different guarantees.

Comment (Alopalao) [reply docs/blueprints/EP038.rst:38]: I see, will explain better on the blueprint. By customer I meant an external connection but I could not make it possible. Changing it to not NNI. So `coloring` will create flows only for discovered NNIs.

Comment (viniarck) [reply docs/blueprints/EP038.rst:38]: I see. Right, if it `coloring` keeps going hand in hand with `of_lldp` and the discovered NNIs that facilitates.

**Comment by viniarck** [comment]
@Alopalao let me know when you need another review, the updated text look good, looks like these points are still open though:

- Do we really need to manage the general dl_src actions drop? Since flows will be matched with `in_port` too on NNIs? Maybe we don't even need to manage extra drop flows and let the switch itself drop. Unless for sanity sake we'd like to make that explicitly. I'll leave that up to you and @italovalcy to discuss if you really want this to be explicit. 
- How flows will be migrated? Are you going to provide a DB script?
- Partly related: `coloring` is also exposing the `COLOR_FIELD='dl_src'` to be configurable, since that flow is never persisted, if it changes it might also leave flows behind, which probably wouldn't impact but still leftovers, maybe it might be worth documenting in the README if it changes then network operators are supposed to delete the old flows?

File: docs/blueprints/EP038.rst
Comment (viniarck) [inline docs/blueprints/EP038.rst:46]: The mask/wildcard idea is great to minimize the number of flows. Although currently, depending on a dpid value, for instance `cc:4e:24:4b:11:11:11:11`, [it can fill up the entire 6 bytes space without needing any of the 'ee' fill placeholder bytes](https://github.com/kytos-ng/coloring/blob/master/main.py#L130-L132), if we'll use mask:

- We'll need to reserve more bits just so we can be sure that wilcard will match accordingly
- We'll need to make sure and document which [`COLOR_FIELD`](https://github.com/kytos-ng/coloring/blob/master/settings.py#L3) will actually be supported and tested (only `dl_src` and `dl_dst` or should the solution also work for the other ones)?

Could you double check and cover this too?

File: docs/blueprints/EP038.rst
Comment (viniarck) [inline docs/blueprints/EP038.rst:26]: Just a heads up that currently [`COLOR_FIELD` is configurable](https://github.com/kytos-ng/coloring/blob/master/main.py#L41), and it's even allowing [`in_port` to be used](https://github.com/kytos-ng/coloring/blob/master/main.py#L138), it might be a good opportunity to reassess which fields should actually be supported from this point forward

**Review by viniarck — COMMENTED** [review]
Great updates @Alopalao, raised new follow ups though.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:58]: Right, I think it should be fine to expose both priorities in `settings.py`. Also, I see that we are changing `of_lldp` flow priorities to 50000 ([PR](https://github.com/kytos-ng/of_lldp/pull/91)) so flexibility is needed. Contacting @italovalcy  about it.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:26]: We need to delete `in_port` since it will be an addition to all NNI-related flows. Also, this configurable option is not really used. We can get away with just using `dl_src`.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:32]: I didn't notice that I deleted the indentation. I will fix it next commit.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:46]: Checking if we could have a mask/wildcard that could cover all necessary cases. We are deleting `COLOR_FIELD` configurability so `dl_src` could just be a set value.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst]: Adding both priorities for match and drop flows to `settings.py`

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:46]: Adding `dl_src` value to `settings.py` to have a default value `"ee:ee:ee:00:00:00/ff:ff:ff:00:00:00"`. It is enough to cover all current colors and makes it modifiable to be reduced or increased.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:26]: Only using `dl_src` as a mask. We do not need to support all the other fields.

File: docs/blueprints/EP038.rst
Comment (viniarck) [reply docs/blueprints/EP038.rst:46]: @Alopalao right. The idea of exposing completely `DL_MASK` seems too much low level too me. Currently, we're not exposing the `b'ee'` byte filler. The idea of supporting a config here is great, maybe we needs to be exported is `dl_src_reserved_bytes = 3` (or something better named)?, meaning that the user will specify how many bytes are being reserved, then the first x MSB bytes would always be filled with `b'ee'` just so the wildcard would work as expected, and then at the runtime you compute the expected mask when sending sending the flows, what do you think? Let me know if you can see a safer way. Limiting the number of bytes also solves the problem of the dpid using the entire 6 byte space too.

File: docs/blueprints/EP038.rst
Comment (viniarck) [reply docs/blueprints/EP038.rst:26]: Cool. Right, let's officialize this in the v1 and make sure that in the changelog we also explain that we've restricted and broken compatibility to simplify this NApp and keep it easier to maintain.

File: docs/blueprints/EP038.rst
Comment (Alopalao) [reply docs/blueprints/EP038.rst:46]: That looks great. We could name it `DL_MASK_BYTES`. But to confirm I am understanding well. For this variable to have:
- A value 5, means `dl_src = "ee:ee:ee:ee:ee:00/ff:ff:ff:ff:ff:00"
- A value 2, means `dl_src = "ee:ee:00:00:00:00/ff:ff:00:00:00:00"

Is this correct?

Comment (viniarck) [reply docs/blueprints/EP038.rst:46]: Yes, @Alopalao. Great variable name `DL_MASK_BYTES` by the way. `6` - `DL_MASK_BYTES` represents how many bytes would be available for coloring each (dpid) neighbor, we need also a default value, probably 3 bytes (24 bits) is reasonable, especially since in the uplinks `in_port` will also be used. 

It's worth also confirming that indeed if a neighbor dpid that shares the same last 3 dpid bytes wouldn't cause issues (prior to this change each neighbor was guaranteed to have a different color since it was interpolating the entire 6 bytes). Maybe including the full example would be an opportunity to do so, unless you're crystal clear that it'll work then it's OK too.

File: docs/blueprints/EP038.rst
Comment (italovalcy) [inline docs/blueprints/EP038.rst:90]: @Alopalao please add an additional section on this blueprint to cover other implementation details that will need to be delivered after this blue print. For instance:

- should we provide a migration script to take care of the old flows? 
- what should happen if the operator changes  the DL_MASK on settings? will we send a DELETE all flows with coloring's cookie ID prefix at the begin? (when switch connects?)
- suppose that you want to disable this protection, how should the operator proceed to disable this protection? (maybe something with the PRIORITY_DROP to not install the drop flow -- 0?)

**Review by italovalcy — APPROVED** [review]
Looks good to me! Excellent progress, Aldo!! Just left some comments to collaborate with the discussion!

File: docs/blueprints/EP038.rst
Comment (italovalcy) [reply docs/blueprints/EP038.rst:90]: maybe this new section could be called "Points of attention" or similar.

Just got a new one:

- The color-to-field method on coloring should be adapted to comply with the new dl_src prefix

**Comment by Alopalao** [comment]
This implementation will be on hold for some time until:
- MAC address needs to be changed. Start using addresses reserved from IETF/IANA (`00-00-5E-00-53-00` - `00-00-5E-00-53-FF`)
- Attribute in config file to determine range of assigned MAC addresses
- `coloring` needs a counter to keep track of links.

**Pull request closed without merging**
